### ISSUE TITLE
update after mutation does not trigger UI changes

### ISSUE THREAD

**Issue by FrancescoSaverioZuppichini** [apollographql/apollo-client#6150]
Dear all,

First of all thank you for the amazing library. I have the following problem. Assuming we have to components, `A` and `B`. Component B uses a query `GET_ME` to get the user's information. They have the following form:

```javascript
// result of query GET_ME
me : {

....
decks: [...] // array of objects
}
```

Component `B` is 

```javascript
// component B
export default function B() {
    const { error, data } = useQuery(queries.GET_ME)
    return (
      // show data.me.decks
)
```

Very easy, component B just show the `deck` array that is inside the object returned from the `GET_ME` query.

Now, in the component `A` we want to remove on of the item of this array (`me.decks`).

```javascript
component A
export default function A({ id }) {
    const [deleteDeck, { deleteDeckError }] = useMutation(mutations.DELETE_DECK, {
        onCompleted({ deleteDeck }) {
            console.log(deleteDeck)
        },
        update(cache, { data: { deleteDeck } }) {
            let { me } = cache.readQuery({ query: queries.GET_ME })
            console.log(me.decks.length)
            me.decks = me.decks.filter(deck => deck.id !== deleteDeck.id)
            console.log(me.decks.length)
            cache.writeQuery({
                query: queries.GET_ME,
                data: me,
            })
        }
    })

```

Here we are just using the `DELETE_DECK` mutation to remove a deck from a user. The code works perfectly, the correct document is removed from the server **but** the UI in component B is not update!. In the `update` function in the mutation inside component `B` I am updating the local state of the cache. Assuming I have only one item inside `me.decks`, the output of the console logs is

```
1
0
```

Showing that the item was correctly removed, still no changes in the UI of component `B`. These two components do not share a relationship, (one if not the parent of the other)

What I am doing wrong?

Thank you.

Bests,

Francesco Saverio

**Intended outcome:**

The UI in component B should be updated

**Actual outcome:**
The UI  in component B is not updated

**Versions**

```
 System:
    OS: Linux 5.3 Ubuntu 18.04.4 LTS (Bionic Beaver)
  Binaries:
    Node: 12.13.0 - /usr/local/bin/node
    npm: 6.12.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 80.0.3987.149
    Firefox: 75.0
  npmPackages:
    @apollo/react-hooks: ^3.1.4 => 3.1.4 
    apollo-boost: ^0.4.7 => 0.4.7 
```

**Comment by jspizziri** [comment]
## TL;DR;

I'd recommend double-checking that there is actually a _change_ in the data that you're writing to the cache, as if there isn't one, Apollo won't trigger an update in your component. In my case, this was the issue.

## Here We Go

FWIW, I was having a problem that displayed identical symptoms. What it boiled down to for me was effectively 2 things:

1. No updates were being triggered because whatever change detection that Apollo does on `writeQuery` were failing.
2. Change detections were failing because I was improperly managing the cache `before` my write.

**Example**

I had a nested data structure that looked something like the following:

```ts
interface Bar {
  id: number;
  isEnabled: boolean;
}

interface Foo {
  id: number;
  bars: Bar[];
}

const foo: Foo = {
  bars: [
    { id: 1, enabled: false },
  ] as Bar[]
}
```

Then I'd query for the foo object and rely on mutations of `Bar` and cache updates to rerender anything that depended on `Foo`. E.g.:

```ts
...
const {loading, error data} = useQuery(GET_FOO, { variables: { id: 1 } });
...
```

Deeper down in the component stack I'd run a mutation on Bar:

```ts
useMutation(DISABLE_BAR, {
  update: (cache: DataProxy, result: FetchResult) => {
    const cachedQuery: any = cache.readQuery({ query: GET_FOO, variables: { id: fooId } });
    ...
    const barIdx = foo.bars.findIndex((bar: any) => bar.id === resp.id)

    if (barIdx >= 0) {
      console.log('bar before modification', foo.bars[barIdx]);
      foo.bars[barIdx].enabled = resp.enabled;
      console.log('bar after modification', foo.bars[barIdx]);
      cache.writeQuery({
        query: GET_FOO,
        variables: { id: fooId } ,
        data: { foo },
      });
    }
  },
});
```

However in the parent level component nothing was rerendering. Then I added logging (seen above in the mutation) to check the data fetched from the cache, and the data I was writing to the cache and realized that `enabled` was identical both before and after setting `foo.bars[barIdx].enabled = resp.enabled`. So basically, nothing was changing, so Apollo was just doing its job. I then realized that elsewhere I was improperly managing the cache which was resulting in the `enabled` property to get out of sync with the canonical data.

In hindsight, mine was a simple problem, but depending on the complexity of the application one that could be an easy oversight. Not sure if this is the same problem everyone else is experiencing, but hopefully it helps.

**Comment by hwillson** [comment]
It doesn't sound like there is an outstanding issue here, so closing. Thanks!
